### PR TITLE
[snappy] Update to 1.1.10 and disable optimizations based on hardware support

### LINKED ISSUE #14520: [snappy] Please do NOT build snappy using special processor instructions (BMI2, SSSE3)
****

**Is your feature request related to a problem? Please describe.**
Compiling snappy using processor instructions available in building computer makes the produced programs undeployable to customers computers. Please never enable special instructions to vcpkg ports and make them optional if possible.

**Proposed solution**
SNAPPY_HAVE_BMI2, SNAPPY_HAVE_SSSE3 should be 0 while building

**Describe alternatives you've considered**
I am not aware if there are other options for vcpkg users to interfere with port building. Please advise.

### PR DESCRIPTION
**Pull request by Osyotr**
Fixes #14520
Continuation of #25574

### CODE REVIEW

File: ports/snappy/portfile.cmake [hunk 1]
```diff
@@ -8,9 +8,16 @@ vcpkg_from_github(
         fix_clang-cl_build.patch
 )
 
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+        bmi2  SNAPPY_HAVE_BMI2
+        ssse3 SNAPPY_HAVE_SSSE3
+)
+
```
Comment (Neumann-A) [inline ports/snappy/portfile.cmake:16]: hmm no. This uses `check_cxx_source_compiles` to check if those are available so providing a toolchain with the correct flags will take care of this.

**Comment by Neumann-A** [comment]
you might want to fix https://github.com/google/snappy/blob/af720f9a3b2c831f173b6074961737516f2d3a46/CMakeLists.txt#L113-L115 instead

File: ports/snappy/portfile.cmake (hunk 1 above)
Comment (Osyotr) [reply ports/snappy/portfile.cmake:16]: You mean
```cmake
if("${PORT}" STREQUAL "snappy")
    set(SNAPPY_HAVE_BMI2 OFF) # or ON
    set(SNAPPY_HAVE_SSSE3 OFF) # or ON
endif()
```
in triplet file? IIRC some ports have AVX feature, are they also incorrect?

Comment (Neumann-A) [reply ports/snappy/portfile.cmake:16]: AVX should come from the toolchain unless somebody builds for all archs and does a runtime arch selection.

```
if("${PORT}" STREQUAL "snappy")
    list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS -DSNAPPY_HAVE_BMI2=OFF -DSNAPPY_HAVE_SSSE3=OFF)
endif()
```

**Comment by Osyotr** [comment]
> you might want to fix https://github.com/google/snappy/blob/af720f9a3b2c831f173b6074961737516f2d3a46/CMakeLists.txt#L113-L115 instead

These variables are used only when `SNAPPY_BUILD_TESTS` is enabled, no action needed.

**Comment by Neumann-A** [comment]
> These variables are used only when SNAPPY_BUILD_TESTS is enabled, no action needed.

Ok was wondering why nobody noticed that yet ;)

**Comment by Neumann-A** [comment]
Also this is a problem for MSVC since it allows compiling intrinsics without the required flags. clang-cl will fail the `check_cxx_source_compiles` check without the required flags.

**Comment by LilyWangLL** [comment]
Ping @Osyotr for response. Could you please resolve the conflicts?

**Review by LilyWangLL — DISMISSED** [review]

**Comment by BillyONeal** [comment]
I believe this is tagged vcpkg-team-review because we aren't sure if we want to dig the 'features used to control CPU features' hole deeper. We didn't have time to discuss this week but we'll probably do so next week.

**Comment by BillyONeal** [comment]
Another 'features used to control CPU features' https://github.com/microsoft/vcpkg/pull/29093

**Comment by Osyotr** [comment]
If you don't like these options being features, let's disable them by default. Leaving these options enabled breaks binarycaching.

**Comment by BillyONeal** [comment]
The vcpkg maintainers discussed this on Thursday. We would be happy with the following path forward:

* The BMI2 and SSSE3 behaviors are off by default.
* The port `snappy` looks for `SNAPPY_BMI2' and 'SNAPPY_SSSE3' variables to be set before `portfile.cmake` starts running. These variables are intended to be set by the triplet. This goes for all other similar CPU feature tests in the future ('PORTNAME_CPUTHING') and allows a way for progress to be made here.

Note that while we have become less happy with 'half measures' like this over time, we don't expect to make meaningful progress on defining what the 'total matrix of triplet variables to control all the CPU things' is going to look like over time. Moreover, this path does not close the door on introducing a hypothetical `VCPKG_BMI2` or similar triplet variable should we choose to add something like that in the future, as a new version of the port could look for either a port-specific setting or a vcpkg-wide setting, if applicable.

We are still in agreement that port features should not be used to control CPU features as there is no way to require that a feature is off.

**Review by LilyWangLL — APPROVED** [review]

**Comment by BillyONeal** [comment]
That resolution is even better, thanks!

**Pull request merged**